sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java

```diff
@@ -0,0 +1,456 @@
+    addTimedEvent(TimedEvent.create(clock.nowNanos(), name, attributes));
+  }
+
+  @Override
+  public void addEvent(Event event) {
+    addTimedEvent(TimedEvent.create(clock.nowNanos(), event));
+  }
+
+  private void addTimedEvent(TimedEvent timedEvent) {
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling addEvent() on an ended Span.");
+        return;
+      }
+      getInitializedEvents().add(timedEvent);
+      totalRecordedEvents++;
+    }
+  }
+
+  @Override
+  public void addLink(Link link) {
+    Preconditions.checkNotNull(link, "link");
+    addLinkInternal(SpanData.Link.create(link.getContext(), link.getAttributes()));
+  }
+
+  @Override
+  public void addLink(SpanContext spanContext) {
+    addLinkInternal(SpanData.Link.create(spanContext));
+  }
+
+  @Override
+  public void addLink(SpanContext spanContext, Map<String, AttributeValue> attributes) {
+    addLinkInternal(SpanData.Link.create(spanContext, attributes));
+  }
+
+  private void addLinkInternal(SpanData.Link link) {
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling addLink() on an ended Span.");
+        return;
+      }
+      getInitializedLinks().add(link);
+      totalRecordedLinks++;
+    }
+  }
+
+  @Override
+  public void setStatus(Status status) {
+    Preconditions.checkNotNull(status, "status");
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling setStatus() on an ended Span.");
+        return;
+      }
+      this.status = status;
+    }
+  }
+
+  @Override
+  public void updateName(String name) {
+    Preconditions.checkNotNull(name, "name");
+    synchronized (this) {
+      this.name = name;
+    }
+  }
+
+  @Override
+  public void end() {
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling end() on an ended Span.");
+        return;
+      }
+      endNanoTime = clock.nowNanos();
+      hasBeenEnded = true;
+    }
+    spanProcessor.onEndSync(this);
+  }
+
+  @Override
+  public SpanContext getContext() {
+    return context;
+  }
+
+  @Override
+  public boolean isRecordingEvents() {
+    return true;
+  }
+
+  void addChild() {
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling end() on an ended Span.");
+        return;
+      }
+      numberOfChildren++;
+    }
+  }
+
+  @GuardedBy("this")
+  private AttributesWithCapacity getInitializedAttributes() {
+    if (attributes == null) {
+      attributes = new AttributesWithCapacity(traceConfig.getMaxNumberOfAttributes());
+    }
+    return attributes;
+  }
+
+  @GuardedBy("this")
+  private EvictingQueue<TimedEvent> getInitializedEvents() {
+    if (events == null) {
+      events = EvictingQueue.<TimedEvent>create((int) traceConfig.getMaxNumberOfEvents());
+    }
+    return events;
+  }
+
+  @GuardedBy("this")
+  private EvictingQueue<SpanData.Link> getInitializedLinks() {
```

Comment:
Why do we not use Link interface directly here?

Reply:
Because `Link` is an interface and we need a concrete implementation of it to implement `addLink(SpanContext spanContext)` and `addLink(SpanContext spanContext, Map<String, AttributeValue> attributes)`. If we don't use `SpanData.Link` here, we need to add a `LinkImpl` in the SDK. WDYT?

Reply:
But the list of Links can still be the interface. When we add to the list we add the SpanData.Link

Reply:
I got your point - updated in https://github.com/open-telemetry/opentelemetry-java/commit/d3ac6dc171f04a8e5cea0b4995134c7fc8e8fd36.